File: go/security_test.go
package bench_test

import (
	"context"
	bench "github.com/trybench/bench-sdk/go"
	"io"
	"net/http"
	"strings"
	"testing"
	"time"
)

type privacyTransport func(*http.Request) (*http.Response, error)

func (f privacyTransport) RoundTrip(r *http.Request) (*http.Response, error) { return f(r) }

func TestPrivacyEnvelopeAndBoundedText(t *testing.T) {
	var body string
	client, err := bench.New(bench.Options{APIKey: "bench_sk_synthetic_test_key", Repository: "fixture/go", Branch: "main", SystemName: "{\"email\":\"person@example.test\",\"password\":\"SYNTHETIC_PRIVATE_VALUE\",\"secret\":\"bench_sk_NOT_A_REAL_SECRET\"}", CaptureContent: true, Transport: privacyTransport(func(r *http.Request) (*http.Response, error) {
		b, _ := io.ReadAll(r.Body)
		body = string(b)
		return &http.Response{StatusCode: 201, Body: io.NopCloser(strings.NewReader("{}")), Header: make(http.Header)}, nil
	})})
	if err != nil {
		t.Fatal(err)
	}
	_, _ = bench.Trace(context.Background(), client, bench.SpanInput{Name: "privacy", Input: strings.Repeat("x", 1000000)}, func(context.Context) (bool, error) { return true, nil })
	client.Shutdown(context.Background())
	if strings.Contains(body, "person@example.test") || strings.Contains(body, "NOT_A_REAL_SECRET") || !strings.Contains(body, "[CONTENT_LIMIT]") || strings.Contains(body, "SYNTHETIC_PRIVATE_VALUE") {
		t.Fatal("privacy policy was not applied to envelope/content")
	}
	for _, branch := range []string{"person@example.test", "bench_sk_NOT_A_REAL_SECRET", "bad\nbranch"} {
		if _, err := bench.New(bench.Options{APIKey: "bench_sk_synthetic_test_key", Repository: "fixture/go", Branch: branch}); err == nil {
			t.Fatal("unsafe identity accepted")
		}
	}
}

func TestConcurrentFlushAndShutdownHonorCallerDeadline(t *testing.T) {
	entered, release, finished := make(chan struct{}), make(chan struct{}), make(chan struct{})
	client, err := bench.New(bench.Options{APIKey: "bench_sk_synthetic_test_key", Repository: "fixture/go", Branch: "main", Transport: privacyTransport(func(r *http.Request) (*http.Response, error) {
		close(entered)
		<-release
		return &http.Response{StatusCode: 201, Body: io.NopCloser(strings.NewReader("{}")), Header: make(http.Header)}, nil
	})})
	if err != nil {
		t.Fatal(err)
	}
	_, _ = bench.Trace(context.Background(), client, bench.SpanInput{Name: "request"}, func(context.Context) (bool, error) { return true, nil })
	go func() { client.Flush(context.Background()); close(finished) }()
	<-entered
	defer func() { close(release); <-finished }()
	for _, flush := range []func(context.Context){client.Flush, client.Shutdown} {
		ctx, cancel := context.WithTimeout(context.Background(), 20*time.Millisecond)
		done := make(chan struct{})
		go func() { flush(ctx); close(done) }()
		select {
		case <-done:
		case <-time.After(250 * time.Millisecond):
			cancel()
			t.Fatal("waiting delivery ignored cancellation")
		}
		cancel()
	}
}

File: test/security.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {Bench} from '../dist/index.js';

test('bounded filtering covers envelope names and refuses unsafe routing identity', async()=>{
 let body;
 const options={apiKey:'bench_sk_synthetic_test_key',repository:'fixture/node',branch:'main',systemName:'person@example.test bench_sk_NOT_A_REAL_SECRET',captureContent:true,fetch:async(_,init)=>{body=JSON.parse(init.body);return new Response('{}',{status:201});}};
 const bench=new Bench(options);
 const start=performance.now();
 await bench.trace({name:'privacy',input:'x'.repeat(1000000)},()=> 'x'.repeat(16000));
 await bench.shutdown();
 assert.ok(performance.now()-start<1000, 'privacy processing must stay bounded');
 assert.ok(!JSON.stringify(body).includes('person@example.test'));
 assert.ok(!JSON.stringify(body).includes('NOT_A_REAL_SECRET'));
 assert.ok(body.traces[0].spans[0].input_value.includes('[CONTENT_LIMIT]'));
 for(const branch of ['person@example.test','bench_sk_NOT_A_REAL_SECRET','bad\nbranch'])assert.throws(()=>new Bench({...options,branch}));
});
